### ISSUE TITLE
Client ID, IG_SIG_KEY, SIG_KEY_VERSION

### ISSUE THREAD

**Issue by ranman034** [LevPasha/Instagram-API-python#40]
I am still having problems getting my test.py to run successfully.   Where do I enter my Client ID from Instagram? What are the IG_SIG_KEY and SIG_KEY_VERSION? Do I need to generate those values?  Where in the InstagramAPI.py file do you have to modify for it to work?

**Comment by 179skr** [comment]
If you look in InstagramAPI.py youll find these lines 

USER_AGENT = 'Instagram 9.2.0 Android ({android_version}/{android_release}; 320dpi; 720x1280; {manufacturer}; {model}; armani; qcom; en_US)'.format(**DEVICE_SETTINTS)
    IG_SIG_KEY = '012a54f51c49aa8c5c322416ab1410909add32c966bbaa0fe3dc58ac43fd7ede'
    EXPERIMENTS = 'ig_android_progressive_jpeg,ig_creation_growth_holdout,ig_android_report_and_hide,ig_android_new_browser,ig_android_enable_share_to_whatsapp,ig_android_direct_drawing_in_quick_cam_universe,ig_android_huawei_app_badging,ig_android_universe_video_production,ig_android_asus_app_badging,ig_android_direct_plus_button,ig_android_ads_heatmap_overlay_universe,ig_android_http_stack_experiment_2016,ig_android_infinite_scrolling,ig_fbns_blocked,ig_android_white_out_universe,ig_android_full_people_card_in_user_list,ig_android_post_auto_retry_v7_21,ig_fbns_push,ig_android_feed_pill,ig_android_profile_link_iab,ig_explore_v3_us_holdout,ig_android_histogram_reporter,ig_android_anrwatchdog,ig_android_search_client_matching,ig_android_high_res_upload_2,ig_android_new_browser_pre_kitkat,ig_android_2fac,ig_android_grid_video_icon,ig_android_white_camera_universe,ig_android_disable_chroma_subsampling,ig_android_share_spinner,ig_android_explore_people_feed_icon,ig_explore_v3_android_universe,ig_android_media_favorites,ig_android_nux_holdout,ig_android_search_null_state,ig_android_react_native_notification_setting,ig_android_ads_indicator_change_universe,ig_android_video_loading_behavior,ig_android_black_camera_tab,liger_instagram_android_univ,ig_explore_v3_internal,ig_android_direct_emoji_picker,ig_android_prefetch_explore_delay_time,ig_android_business_insights_qe,ig_android_direct_media_size,ig_android_enable_client_share,ig_android_promoted_posts,ig_android_app_badging_holdout,ig_android_ads_cta_universe,ig_android_mini_inbox_2,ig_android_feed_reshare_button_nux,ig_android_boomerang_feed_attribution,ig_android_fbinvite_qe,ig_fbns_shared,ig_android_direct_full_width_media,ig_android_hscroll_profile_chaining,ig_android_feed_unit_footer,ig_android_media_tighten_space,ig_android_private_follow_request,ig_android_inline_gallery_backoff_hours_universe,ig_android_direct_thread_ui_rewrite,ig_android_rendering_controls,ig_android_ads_full_width_cta_universe,ig_video_max_duration_qe_preuniverse,ig_android_prefetch_explore_expire_time,ig_timestamp_public_test,ig_android_profile,ig_android_dv2_consistent_http_realtime_response,ig_android_enable_share_to_messenger,ig_explore_v3,ig_ranking_following,ig_android_pending_request_search_bar,ig_android_feed_ufi_redesign,ig_android_video_pause_logging_fix,ig_android_default_folder_to_camera,ig_android_video_stitching_7_23,ig_android_profanity_filter,ig_android_business_profile_qe,ig_android_search,ig_android_boomerang_entry,ig_android_inline_gallery_universe,ig_android_ads_overlay_design_universe,ig_android_options_app_invite,ig_android_view_count_decouple_likes_universe,ig_android_periodic_analytics_upload_v2,ig_android_feed_unit_hscroll_auto_advance,ig_peek_profile_photo_universe,ig_android_ads_holdout_universe,ig_android_prefetch_explore,ig_android_direct_bubble_icon,ig_video_use_sve_universe,ig_android_inline_gallery_no_backoff_on_launch_universe,ig_android_image_cache_multi_queue,ig_android_camera_nux,ig_android_immersive_viewer,ig_android_dense_feed_unit_cards,ig_android_sqlite_dev,ig_android_exoplayer,ig_android_add_to_last_post,ig_android_direct_public_threads,ig_android_prefetch_venue_in_composer,ig_android_bigger_share_button,ig_android_dv2_realtime_private_share,ig_android_non_square_first,ig_android_video_interleaved_v2,ig_android_follow_search_bar,ig_android_last_edits,ig_android_video_download_logging,ig_android_ads_loop_count_universe,ig_android_swipeable_filters_blacklist,ig_android_boomerang_layout_white_out_universe,ig_android_ads_carousel_multi_row_universe,ig_android_mentions_invite_v2,ig_android_direct_mention_qe,ig_android_following_follower_social_context'
    SIG_KEY_VERSION = '4'

**Comment by ranman034** [comment]
@179skr that does not answer my question.  I was wondering if I have to modify anything in the InstagramAPI.py file for it work.

**Comment by 179skr** [comment]
Sorry, no you do not need to modify anything. Im currently using this api without modding and it works.

**Comment by ahmdrz** [comment]
It's working for me too , I haven't any problem

**Comment by ranman034** [comment]
@ahmdrz @179skr Did you have to change anything in the test.py file or other files?

**Comment by ahmdrz** [comment]
@ranman034 No , I didn't change anything in test.py

**Comment by ahmdrz** [comment]
@ranman034 Dude ! this is not official instagram api , so you don't need Client ID

this library work similar with android version , so you need just your username and password to login

**Comment by sosico** [comment]
I'm having issues using the test.py as well I'm getting the following error:
Instagram-API-python-master/InstagramAPI.py", line 475, in generateSignature
    return 'ig_sig_key_version=' + self.SIG_KEY_VERSION + '&signed_body=' + hmac.new(self.IG_SIG_KEY.encode('utf-8'), data.encode('utf-8'), hashlib.sha256).hexdigest() + '.' + urllib.parse.quote(data)
AttributeError: 'module' object has no attribute 'parse'

**Comment by sosico** [comment]
It's a problem with python 2.7. If you are using 2.7 you must use urllib.quote(data) if you use python 3 you need to use urllib.parse.quote(data)

**Comment by jeflopo** [comment]
I'm getting an error response with login endpoint `https://i.instagram.com/api/v1/accounts/login/`

> '{"status": "fail", "message": "Your version of Instagram is out of date. Please upgrade your app to log in to Instagram."}'

The code has 9.2.0 version, my mobile has 9.5.0 version and I've been able to download a 9.7.0 apk

I think that `USER_AGENT`, `IG_SIG_KEY`, `SIG_KEY_VERSION` variables from https://github.com/LevPasha/Instagram-API-python/blob/master/InstagramAPI.py must be updated in order to make this project to work again. Correct me If I'm wrong.

I'm trying to extract the `IG_SIG_KEY` using `gikdbg.art` assembler debugger. I've created an `Android AVD 4.4.2 API lvl 19` using ARM runtime. I have attached it to the debugger, I loaded the module libstrings.so And I found in the source the `GetSignatureString` Method. There's another comment in the assembler code that points to something called `.strlen` And tried some breakpoints / "Follow in dump" in the registers. But I haven't been able to get the key.

**Comment by ahmdrz** [comment]
@jeflopo Special Thanks , Perfect ....
Did you check Instagram-Api for PHP ?

**Comment by ahmdrz** [comment]
@jeflopo Try this : fc4720e1bf9d79463f62608c86fbddd374cc71bbfb98216b52e3f75333bd130d

**Comment by jeflopo** [comment]
@ahmdrz
> Did you check Instagram-Api for PHP ?

I Checked the date of the commits to see if the relevant files were update in the python port. And in the PHP repo.

> Try this : fc4720e1bf9d79463f62608c86fbddd374cc71bbfb98216b52e3f75333bd130d

I Tried It. Same error. This is the IG_SIG_KEY from [Instagram-API (PHP)](https://github.com/mgp25/Instagram-API) 9.40, [Instagram-API-python](https://github.com/LevPasha/Instagram-API-python) has 9.2.0

As I stated above it is possible to download 9.7.0 (but from google play I get 9.6.7, weird..)

**Comment by ahmdrz** [comment]
@jeflopo 

>I Checked the date of the commits to see if the relevant files were update in the python port. And in the PHP repo.

The dates is not important , Because maybe the commits not changed the key.

***

I haven't python on this computer that I used , But I have Go ! I am in travel right now.

I tried this :
```go
package main

import (
   "fmt"

   "github.com/ahmdrz/goinsta"
)

func main() {
    insta := goinsta.New("###", "###")

    if err := insta.Login(); err != nil {
        panic(err)
    }

    defer insta.Logout()

    fmt.Println(insta.GetProfileData())
}
```

And everything is OK ! 

goinsta key : [55e91155636eaa89ba5ed619eb4645a4daf1103f2161dbfe6fd94d5ea7716095](https://github.com/ahmdrz/goinsta/blob/master/goinsta.go#L82)

Instagram-API-python key : [012a54f51c49aa8c5c322416ab1410909add32c966bbaa0fe3dc58ac43fd7ede](https://github.com/LevPasha/Instagram-API-python/blob/master/InstagramAPI.py#L23)

Try this : 55e91155636eaa89ba5ed619eb4645a4daf1103f2161dbfe6fd94d5ea7716095

**Comment by jeflopo** [comment]
My apologies. It was only my mistake with customized code. xd

Good News are that. I can Confirm Instagram-API-python Works with its current key: [012a54f51c49aa8c5c322416ab1410909add32c966bbaa0fe3dc58ac43fd7ede](https://github.com/LevPasha/Instagram-API-python/blob/master/InstagramAPI.py#L23)

I've just made a photo post with Instagram-API-Python

Thanks for the help @ahmdrz And... I didn't knew about goinsta :) I'll give it a try too !

**Comment by h-2-0** [comment]
@jeflopo , @ahmdrz 

Hello! Could you please give me some directions/ideas on how to login as I were using my mobile but from my local machine? The reason why I want to do this is because I keep getting "checkpoint_challenge_required" and I wanted to emulate my mobile from localhost. 

What I have done is I've sniffed my User Agent out of request using Charles Proxy.  Then in InstagramAPI.py I changed the USER_AGENT accordingly but now when I try to log in I get the "Your version is out of date" error message. Is it because a different user agent requires a different signature key? Or the signature key depends on the Instagram Version? Or it depends on the Android version? In any case, how do I know which signature key should I use ?

`USER_AGENT = "Instagram 10.2.1 Android (19/4.4.2; 240dpi; 480x800;...en_US}"`   

Thanks in advance!

**Comment by ahmdrz** [comment]
Hi @h-2-0  did you change IG_SIG_KEY with the latest ?

**Comment by h-2-0** [comment]
Hello @ahmdrz 
I'm using the following setting:
`IG_SIG_KEY = '012a54f51c49aa8c5c322416ab1410909add32c966bbaa0fe3dc58ac43fd7ede'`

This is my work flow
1. Create a instagram account from the web browser. 
2. Run the instagram account using InstagramAPI.py from my local server to like a max of 1000 media at a rate <350 likes/hour. 
3. Next day OR after a few hundreds likes while trying to log in to the account via InstagraAPI  I get  a 'checkpoint required' error message. BUT if I log in from my mobile into the IG account I start a verification process where I have to submit  a code received in my mobile as a text message. Once I submit the code I can enter my account from my mobile.
4. So I try to log in again into my IG account from local server but I get again the 'checkpoint required' error message.  

So basically what I want to do is emulate my mobile in local server so Instagram server won't trigger the checkpoint verification.

**Comment by ahmdrz** [comment]
Ok @h-2-0 , This is instagram security.
Login with instagram app and click on "yes , this is me" and try again.

**Comment by h-2-0** [comment]
@ahmdrz 

I do that with my phone but I still get the checkpoint  once I try to log in from local server. :/

**Comment by ahmdrz** [comment]
@h-2-0 I don't know honestly.
You can ask @mgp25.

**Comment by h-2-0** [comment]
@ahmdrz 

I'm scared of asking @mgp25 ! xD But I'll try to approach them somehow.

By the way here's how they (PHP people) overcome the checkpoint:

[checkpoint_required #379 at mgp25/Instagram-API](https://github.com/mgp25/Instagram-API/issues/379)

Since I don't know PHP a good start for me would be to translate the PHP code into Python. Even if it is translated into pseudo code would be a great help for me.

Thanks @ahmdrz anyway!

**Comment by ahmdrz** [comment]
@h-2-0 I don't know php enough.
But if I found a way i will tell you ?

**Comment by h-2-0** [comment]
@ahmdrz 

Yes, please! \o/ Thanks!

**Comment by h-2-0** [comment]
@ahmdrz 

Few minutes ago I tried to kickstart a conversation at mpg25/InstagramAPI  and unsurprisingly @mgp25 sent the conversation to hell! :+1: See? This is why I prefer to anybody else except @mgp25! xD

Anyway, this is what I have tried so far based on the Checkpoint.php commit:

```
import requests
from InstagramAPI import InstagramAPI

InstaAPI = InstagramAPI(instagram_username, instagram_password)
InstaAPI.login()
login = InstaAPI.LastJson
token = InstaAPI.LastResponse.cookies['csrftoken']

headers = {
'Origin': 'https://i.instagram.com',
'Connection': 'keep-alive',
'Proxy-Connection': 'keep-alive',
'Accept-Language': 'en-US',
'Referer': 'https://i.instagram.com/challenge/4289996693/RT5hdfY1Fw/',
'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
'X-CSRFToken': token,
}
post_request = requests.post(login['checkpoint_url'], headers=headers)
``` 
Unfortunately that gets me a 'Page Not Found'.

I'll post my adventures in trying to port Checkpoint.pho into Python [here](https://github.com/LevPasha/Instagram-API-python/issues/74) thread.

**Comment by mgp25** [comment]
Do you know what the problem is @h-2-0 ? You are trying to do something without knowing/understanding what you are trying to implement.

This is really basic, and is not that i dont like to answer, but i like people to search and struggle, so they can learn, thats the point, to learn.

Lets explain this easy. Checkpoint is just a verification and it consists in a couple of requests. First one is done to take the CSRFToken, this is very easy to do in python, in just 5 lines.

The second request is just a POST request, with static-known fields, its the same as above but with post data, adding one line to the previous code and thats it!

And the third request, guess what? Another POST request! This time it includes the verification code that you should receive via email or SMS

In conclusion, CheckPoint is like, mm.. 20 lines or less? Now that you have more clues, coding it will take you more or less 3 minutes. You can take Checkpoint.php from my repo as reference

Regards and happy xmas

**Comment by ahmdrz** [comment]
@mgp25 
> but i like people to search and struggle so they can learn, thats the point, to learn. :+1:

**Comment by h-2-0** [comment]
@mgp25, 

Aww, thanks for taking the time to explain the whole process in very simple steps easy to understand. I didn't really fully grasp it because, first of all, I didn't had understand the signature key part as I was trying to connect using an incompatible Instagram version/signature key.

My apologies if I annoyed you. I feel like crap now. But you are right. It is better for us to struggle and learn. That's actually how I have learnt computer programming so far (my background is business administration). Although I feel that not having the computer fundamentals gets me sometimes very confused and so desperate that I just come to Github's threads expecting a quick solution. I need more patience, research and reason a bit more before asking in.

Happy holidays to you and any body else reading here!

**Comment by mgp25** [comment]
```python
import requests
import re

username_id = '' # you need to set this

FIRST_STEP = 'https://i.instagram.com/integrity/checkpoint/checkpoint_logged_out_main/'+username_id+'/?next=instagram%3A%2F%2Fcheckpoint%2Fdismiss'
UA = 'Mozilla/5.0 (iPhone; CPU iPhone OS 10_1 like Mac OS X) AppleWebKit/601.1.46 (KHTML, like Gecko) Mobile/13G34 Instagram 10.2.0 (iPhone5,2; iPhone OS 10_1; es_ES; es-ES; scale=2.00; 640x1136)'

headers = {
    'User-Agent': UA
}

response = requests.get(FIRST_STEP, headers=headers)

csrftoken = response.cookies['csrftoken']

## Now you can continue, its easy
```

**Comment by h-2-0** [comment]
@mgp25 
Thank you, sir. I'm starting to play with it. :)

**Comment by h-2-0** [comment]
Hola @mgp25! 
Hello people! 

Sorry for coming back to this. I have been working with this the whole day. For some reason I can't pass the second step even though I have tried to replicate the headers sent via mobile sniffing them with Charles Proxy. This is what I have done so far based on what I understand from [ChallengeSMS.php](https://raw.githubusercontent.com/mgp25/Instagram-API/master/src/ChallengeSMS.php): 

**Step 1**
Here I'm going to get the CSRF token needed for the Challenge page even though the target url shown here is not precisely the one I should open (i.e. checkpoint_url, 'https://i.instagram.com/challenge/<username_id>/<some-random-alphanumeric/). 
```
import requests
username_id = '12345678' 
FIRST_STEP = 'https://i.instagram.com/integrity/checkpoint/checkpoint_logged_out_main/'+username_id+'/?next=instagram%3A%2F%2Fcheckpoint%2Fdismiss'
user_agent = 'Instagram 9.6.0 Android (21/5.0.1; 300dpi; 768x1190; LGE/google; Nexus 4; mako; mako; en_US)'
headers = {
    'User-Agent': user_agent,
}
response = requests.get(FIRST_STEP, headers=headers)
csrftoken = response.cookies['csrftoken']
```
**Step 2**
Now I get the Challenge Page. 
```
checkpoint_url = 'https://i.instagram.com/challenge/128931283/jiadusida9/'
payload = {
"csrfmiddlewaretoken": csrftoken,
}
r = requests.get("", headers=headers, data=payload)
csrftoken = r.cookies['csrftoken']
<Response [200]> 
```
**Step 3**
Now I post a valid phone number and I also add headers since only leaving 'User-Agent' doesn't seem to work.
```
headers = {
'Host': 'i.instagram.com',
'Content-Length': "83",
'Connection': 'keep-alive',
'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
'Origin': 'https://i.instagram.com',
'User-Agent': user_agent,
'Content-Type': 'application/x-www-form-urlencoded',
'Referer': checkpoint_url,
'Accept-Encoding': 'gzip,deflate',
'Accept-Language': 'en-GB,en-US;q=0.8',
'X-Requested-With': 'com.instagram.android'
}
payload = {
"csrfmiddlewaretoken": csrftoken,
"phone_number": "+44 1234 123456",
} 
r = requests.post(checkpoint_url, headers=headers, data=payload)
<Response [403]>
```

Surely I'm missing something very obvious but for the life of me I can't see what it is. Also, how come using the same header fields as in my mobile doesn't achieve the same results via local server? Any ideas, pointers would be greatly appreciated! Thanks!

**Comment by MMrR0b0TT** [comment]
Any update?

**Comment by MMrR0b0TT** [comment]
@h-2-0 did you fix it?

**Comment by maxjung96** [comment]
Hi @mgp25 , thanks for always trying to help :)
I understand how the 3 steps work and in principle it's very easy to code them with requests for the new challenge required, where I have to post first to click the button so IG sents the code via email/SMS. I just get an error concerning cookies every time with this code:
`url = "https://i.instagram.com/challenge/XXX/XXX/"

s = requests.Session()
r = s.get(url)

csrftoken = s.cookies.__getitem__("csrftoken")
mid = s.cookies.__getitem__("mid")
rur = s.cookies.__getitem__("rur")

headers = {
    'Host': 'i.instagram.com',
    'Referer': url,
    'X-CSRFToken': csrftoken,
    'Cookie': "csrftoken="+csrftoken+"; mid="+mid+"; rur="+rur
}
payload={"choice": 0}
post_request = s.post(url, data=payload, cookies=r.cookies.get_dict())
print post_request.content`

I had a look at the network traffic when clicking the button to send the code and there are 3 cookies required, which I get with the first GET request and pass them on to the POST request with the choice.
Always in the answer it says:
This page could not be loaded. If you have cookies disabled in your browser, or you are browsing in Private Mode, please try enabling cookies or turning off Private Mode, and then retrying your action.

Does anybody have a hint how it works? 
Thanks, would really help me :)

**Comment by Riyaz7364** [comment]
can anyone fix this issue ??
and anyone here  to help me to create instagram bot in java i cant generate cookies always says Your version of Instagram is out of date. Please upgrade your app to log in to Instagram. i am trying to send llogin request from java to avoid ip block
